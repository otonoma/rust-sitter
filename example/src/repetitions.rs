pub mod grammar {
    use rust_sitter::{Rule, Spanned};

    #[derive(Debug, Rule)]
    #[language]
    #[allow(dead_code)]
    pub struct NumberList {
        #[sep_by1(",")]
        #[leaf(pattern(r"\d+"))]
        numbers: Spanned<Vec<Spanned<i32>>>,
    }

    #[derive(Rule)]
    #[extra]
    struct Whitespace {
        #[leaf(pattern(r"\s"))]
        _whitespace: (),
    }
}

// TODO: Currently not allowed, needs to be fixed.
// pub mod grammar2 {
//     use rust_sitter::{Rule, Spanned};
// 
//     #[derive(Debug, Rule)]
//     #[language]
//     #[allow(dead_code)]
//     pub struct NumberList {
//         #[leaf(pattern(r"\d+"))]
//         numbers: Spanned<Vec<Spanned<i32>>>,
//     }
// 
//     #[derive(Rule)]
//     #[extra]
//     struct Whitespace {
//         #[leaf(pattern(r"\s"))]
//         _whitespace: (),
//     }
// }
// 
// pub mod grammar3 {
//     use rust_sitter::{Rule, Spanned};
// 
//     #[derive(Debug, Rule)]
//     #[language]
//     #[allow(dead_code)]
//     pub struct NumberList {
//         #[sep_by(",")]
//         #[leaf(pattern(r"\d+"))]
//         numbers: Spanned<Vec<Spanned<Option<i32>>>>,
//         #[skip(123)]
//         metadata: u32,
//     }
// 
//     #[derive(Rule)]
//     #[extra]
//     struct Whitespace {
//         #[leaf(pattern(r"\s"))]
//         _whitespace: (),
//     }
// }

#[cfg(test)]
mod tests {
    use super::*;
    use rust_sitter::Language;

    #[test]
    fn repetitions_grammar() {
        // Bug in latest tree-sitter: empty parse on a top-level repeat1 segfaults.
        insta::assert_debug_snapshot!(grammar::NumberList::parse(""));
        insta::assert_debug_snapshot!(grammar::NumberList::parse("1"));
        insta::assert_debug_snapshot!(grammar::NumberList::parse("1, 2"));
    }

    // #[test]
    // fn repetitions_grammar2() {
    //     insta::assert_debug_snapshot!(grammar2::parse(""));
    //     insta::assert_debug_snapshot!(grammar2::parse("1"));
    //     insta::assert_debug_snapshot!(grammar2::parse("1 2"));
    // }

    // #[test]
    // fn repetitions_grammar3() {
    //     insta::assert_debug_snapshot!(grammar3::parse(""));
    //     insta::assert_debug_snapshot!(grammar3::parse("1,"));
    //     insta::assert_debug_snapshot!(grammar3::parse("1, 2"));
    //     insta::assert_debug_snapshot!(grammar3::parse("1,, 2"));
    //     insta::assert_debug_snapshot!(grammar3::parse("1,, 2,"));
    // }
}
