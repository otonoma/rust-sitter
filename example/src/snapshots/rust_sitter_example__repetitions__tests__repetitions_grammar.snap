---
source: example/src/repetitions.rs
expression: "grammar::NumberList::parse(\"\")"
---
ParseResult {
    result: Some(
        NumberList {
            numbers: Spanned {
                value: [],
                position: Position {
                    bytes: 0..0,
                    start: Point {
                        line: 1,
                        column: 1,
                    },
                    end: Point {
                        line: 1,
                        column: 1,
                    },
                },
            },
        },
    ),
    errors: [
        ParseError {
            node_position: Position {
                bytes: 0..0,
                start: Point {
                    line: 1,
                    column: 1,
                },
                end: Point {
                    line: 1,
                    column: 1,
                },
            },
            error_position: Position {
                bytes: 0..0,
                start: Point {
                    line: 1,
                    column: 1,
                },
                end: Point {
                    line: 1,
                    column: 1,
                },
            },
            lookaheads: [],
            reason: Error,
        },
    ],
}
